content/blog/copilot-lessons/index.md

```diff
@@ -0,0 +1,131 @@
+---
+**Artur Laksberg is on the Copilot team here at Pulumi. CoPilot is an AI assistant for cloud infrastructure. Today he shares key lessons from bringing Copilot to production—including an unexpected discovery about AI hallucinations.**
+
+While reviewing user feedback for Pulumi Copilot, our AI assistant for cloud infrastructure, one message caught my eye: "Your tool doesn't know anything!". Having just made some changes, I braced for the worst. But the evals were still looking strong, so what was going on?
+
+<!--more-->
+
+The user was trying to force-delete a stack that still had resources. Copilot confidently suggested using a `--force` flag, which is a perfectly logical solution, except this flag doesn't exist in Pulumi. This was pure hallucination!
+
+Welcome to a day in the life of building with LLMs. But this particular error, this is a specific type of hallucination that I now have a solution for. And now that [we've launched our REST API](/blog/pulumi-copilot-rest/) and Copilot has been running in production for a bit, **I want to share some lessons we've learned building with LLMs**.
+
+Let's start with the fundamental tension between software engineering and prompt engineering.
+
+## Engineering for Reality: Software Engineering vs Prompt Engineering
+
+![Software Engineering vs Prompt Engineering](soft-eng.png)
+
+When building LLM-powered applications, it's tempting to throw every task at the model. Modern LLMs can generate code, format text, and create clickable links. But this approach carries hidden costs.
+
+We learned a key lesson building Copilot: let the LLM do what it does best, and use regular code for everything else.
+
+Take a seemingly simple feature we built early on: listing a user's Pulumi stacks with clickable links based on data from a backend API. Our first implementation used a complex prompt instructing the LLM to construct URLs in the format `app.pulumi.com/org/project/stack`. The prompt explained the format, provided examples, and asked the LLM to generate these links from JSON data it had.
+
+It worked - almost all of the time. But we were seeing occasional malformed URLs and more importantly, we were burning input tokens (and money) with our complicated prompt and having an AI construct strings that could be deterministically generated.
+
+The solution was straightforward: generate the full links in the backend service and include them directly in the context. The LLM then needs no instructions on how to create them. Simple stuff, but it gave us faster responses, perfect URLs, and lower costs.
+
+When you find yourself writing elaborate prompts to handle structured data transformations, stop and ask: Could traditional code do this better? Could this be decomposed so that the LLM did less?
```

Comment:
```suggestion
When you find yourself writing elaborate prompts to handle structured data transformations, stop and ask: Could traditional code do this better? Could this be decomposed so that the LLM does less?
```